### PR TITLE
REF: maybe_promote refactor/cleanup

### PR DESCRIPTION
**Pull request by jbrockmendel**
There is some casting we currently do in many places in this function.  Instead this PR implements `_ensure_dtype_type` and calls it once at the end of the function.

Removes _check_lossless_cast which can be replaced with np.can_cast

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -477,37 +459,48 @@ def maybe_promote(dtype, fill_value=np.nan):
...
-    bool
+    object
     """
-    casted = dtype.type(value)
-    if casted == value:
-        return True
-    return False
+
+    # Start with exceptions in which we do _not_ cast to numpy types
+    if is_extension_array_dtype(dtype):
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: could be elif

**Comment by jbrockmendel** [comment]
rebased+green

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: changed

**Comment by jreback** [comment]
thanks

**Pull request merged**